### PR TITLE
planner: improve row count estimation for index range containing correlated columns

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/9722

### What is changed and how it works?

After appending `col = correlation_col` into access conditions, adjust the estimated row count by multiplying a factor for each condition appended, and the factor is computed as `1 / NDV`.

### Check List 

Tests 

 - Integration test

Code changes

N/A

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: planner/core/logical_plans.go [hunk 1]
```diff
@@ -468,24 +468,31 @@ func (ds *DataSource) deriveIndexPathStats(path *accessPath) (bool, error) {
...
-		} else {
+		if len(accesses) > 0 && ds.statisticTable.Pseudo {
 			path.countAfterAccess = ds.statisticTable.PseudoAvgCountPerValue()
+		} else {
+			for i := range accesses {
+				col := path.idxCols[path.eqCondCount+i]
+				hist, ok := ds.statisticTable.Columns[col.ID]
+				var ndv float64
+				if ok && hist.Count > 0 {
+					factor := float64(ds.statisticTable.Count) / float64(hist.Count)
```
Comment (lamxTyler) [inline planner/core/logical_plans.go]: Why `ds.statisticTable.Count`? This factor does not consider the selectivity of previous index conditions.

File: planner/core/stats.go [hunk 2]
```diff
@@ -85,13 +98,7 @@ func (ds *DataSource) getStatsByFilter(conds expression.CNFExprs) (*property.Sta
...
 	}
 	for i, col := range ds.Columns {
-		hist, ok := ds.statisticTable.Columns[col.ID]
-		if ok && hist.Count > 0 {
-			factor := float64(ds.statisticTable.Count) / float64(hist.Count)
-			profile.Cardinality[i] = float64(hist.NDV) * factor
-		} else {
-			profile.Cardinality[i] = profile.RowCount * distinctFactor
-		}
+		profile.Cardinality[i] = ds.getColumnNDV(col.ID)
```
Comment (zz-jason) [inline planner/core/stats.go]: this seems incorrect, should it be:
```go
hist, ok := ds.statisticTable.Columns[colID]
if ok && hist.Count > 0 {
	profile.Cardinality[i] = float64(hist.Count)
} else {
	profile.Cardinality[i] = ds.statisticTable.Count
}
```
@winoros

Comment (winoros) [reply planner/core/stats.go]: No, `cardinality` stores NDV not the total row count.

Comment (zz-jason) [reply planner/core/stats.go]: it's better to s/Cardinality/NDV/ to eliminate ambiguity

**Pull request merged**